synapse/storage/databases/main/roommember.py

```diff
@@ -1019,7 +1019,17 @@ async def _check_host_room_membership(
 
     @cached(iterable=True, max_entries=10000)
     async def get_current_hosts_in_room(self, room_id: str) -> Set[str]:
-        """Get current hosts in room based on current state."""
+        """
+        Get current hosts in room based on current state.
+
+        The heuristic of sorting by servers who have been in the room the
+        longest is good because they're most likely to have anything we ask
+        about.
+
+        Returns:
+            Returns a list of servers sorted by longest in the room first. (aka.
+            sorted by join with the lowest depth first).
+        """
 
         # First we check if we already have `get_users_in_room` in the cache, as
         # we can just calculate result from that
```

Comment:
I guess this isn't a total drop-in replacement as before since we don't do any sorting for this short-cut grabbing the users from cache. Can we have any assumption on how the `get_users_in_room.cache` is ordered?

Reply:
Updated `get_users_in_room` to return users sorted by lowest -> highest depth. And the `set` comprehension works as expected so duplicates don't change the order.

```py
users = ["foo", "bar", "baz", "foo"]
domains = {u for u in users}

print(f"domains={domains}")
#  -> domains={'foo', 'bar', 'baz'}
```

Reply:
The `set` comprehension actually doesn't maintain the order. Just got lucky when trying this out the first time, see https://github.com/matrix-org/synapse/pull/13575#discussion_r952748930